### PR TITLE
Use MPI_Isend in MpiConsoleReporter to avoid deadlock

### PR DESCRIPTION
**Pull request by starintheuniverse**
## Description
Implements the solution proposed in #655 to a communication bug in MpiConsoleReporter.

### CODE REVIEW

**Comment by onqtam** [comment]
@philbucher @BerengerBerthoul

**Comment by philbucher** [comment]
Cool, thx! 👍 
I can review properly beginning of next week

File: doctest/extensions/mpi_reporter.h [hunk 1]
```diff
@@ -154,6 +155,21 @@ struct MpiConsoleReporter : public ConsoleReporter {
...
+      requests.reserve(m_failure_str_queue.size());  // avoid realloc & copy of MPI_Request
+      for (const std::pair<std::string, int> &failure : m_failure_str_queue)
+      {
+        const std::string & failure_str = failure.first;
+        const int failure_line = failure.second;
+
+        int failure_msg_size = static_cast<int>(failure_str.size());
+
+        requests.push_back(MPI_REQUEST_NULL);
+        MPI_Isend(failure_str.c_str(), failure_msg_size, MPI_BYTE,
```
Comment (philbucher) [inline doctest/extensions/mpi_reporter.h:168]: I think you need to change also the `MPI_Recv` to `MPI_Irecv`

File: doctest/extensions/mpi_reporter.h [hunk 2]
```diff
@@ -196,6 +212,9 @@ struct MpiConsoleReporter : public ConsoleReporter {
           s << "\n";
         }
       }
+
+      MPI_Waitall(int(requests.size()), requests.data(), MPI_STATUSES_IGNORE);
```
Comment (philbucher) [inline doctest/extensions/mpi_reporter.h:216]:
```suggestion
      MPI_Waitall(static_cast<int>(requests.size()), requests.data(), MPI_STATUSES_IGNORE);
```

is the cast even needed?

File: doctest/extensions/mpi_reporter.h [hunk 3]
```diff
@@ -154,6 +155,21 @@ struct MpiConsoleReporter : public ConsoleReporter {
...
       MPI_Comm_rank(MPI_COMM_WORLD, &rank);
 
+      std::vector<MPI_Request> requests;
+      requests.reserve(m_failure_str_queue.size());  // avoid realloc & copy of MPI_Request
+      for (const std::pair<std::string, int> &failure : m_failure_str_queue)
+      {
+        const std::string & failure_str = failure.first;
+        const int failure_line = failure.second;
+
+        int failure_msg_size = static_cast<int>(failure_str.size());
```
Comment (philbucher) [inline doctest/extensions/mpi_reporter.h:165]:
```suggestion
        const int failure_msg_size = static_cast<int>(failure_str.size());
```

File: doctest/extensions/mpi_reporter.h [hunk 4]
```diff
@@ -66,6 +66,7 @@ struct MpiConsoleReporter : public ConsoleReporter {
       return nullStream;
     }
   }
+  std::vector<std::pair<std::string, int>> m_failure_str_queue = {};
```
Comment (philbucher) [inline doctest/extensions/mpi_reporter.h:69]:
```suggestion
  std::vector<std::pair<std::string, int>> m_failure_str_queue;
```

I think this is not needed (?)

**Review by philbucher — COMMENTED** [review]
Thanks for the fix!
Overall the proposed solution looks good to me, you are right that we should not rely on the implementation details of some MPI implementations
`MPI_Send` is usually blocking until a matching receive is issued, I guess this was not noticed so far due to small message sizes, which were buffered

I made minor comments, mostly style, but also about the now not matching `MPI_Irecv`

BTW I checked `MPI_Bsend` (which is a buffered send), but found it to be inconvenient to use since one needs to attach a buffer manually

File: doctest/extensions/mpi_reporter.h (hunk 2 above)
Comment (BerengerBerthoul) [reply doctest/extensions/mpi_reporter.h:216]: yes you don't need it, it will be converted anyway

**Comment by BerengerBerthoul** [comment]
Thanks @starintheuniverse for the PR and @philbucher for the review. It looks good to me

File: doctest/extensions/mpi_reporter.h (hunk 4 above)
Comment (starintheuniverse) [reply doctest/extensions/mpi_reporter.h:69]: True, but I think it is more consistent with the member declarations seen in other classes in the file.

**Comment by starintheuniverse** [comment]
Thanks @philbucher for the review and also @BerengerBerthoul for taking a look!

**Comment by starintheuniverse** [comment]
I just noticed a separate preexisting bug in `MpiConsoleReporter` that I'd like to fix before this gets merged: Program will hang if `WARN` is used inside an `MPI_TEST_CASE`.

It seems that `st.numAssertsFailedCurrentTest` does not count `WARN` asserts, so does not equal the number of messages logged and sent. There will be a mismatch between number of messages sent and received. An easy fix is to `MPI_Reduce` over the actual number of messages rather than the number of failed asserts. This is particularly easy now that `MPI_Isend` is being used in the same function as the `MPI_Reduce`.

**Comment by philbucher** [comment]
no more comments from my side, please go ahead :)
Thanks again!

**Comment by Saalvage** [comment]
Thanks for your contribution :)

**Pull request merged**